micropipenv.py

```diff
@@ -295,6 +295,76 @@ def _compute_pipfile_hash(pipfile):  # type: (Dict[str, Any]) -> str
     return hashlib.sha256(content.encode("utf8")).hexdigest()
 
 
+def _compute_poetry_hash(pyproject):  # type: (MutableMapping[str, Any]) -> str
+    """Compute pyproject.toml hash based on poetry content."""
+    poetry_data = pyproject["tool"]["poetry"]
+    relevant_keys = ["dependencies", "dev-dependencies", "source", "extras"]
+    relevant_content = {key: poetry_data.get(key) for key in relevant_keys}
+
+    return hashlib.sha256(json.dumps(relevant_content, sort_keys=True).encode()).hexdigest()
+
+
+def _get_installed_python_version():  # type: () -> str
+    return "{}.{}".format(sys.version_info.major, sys.version_info.minor)
+
+
+def _validate_poetry_python_version(
+    poetry_lock, current_python, debug=False
+):  # type: (MutableMapping[str, Any], str, bool) -> None
+    # TODO: Implement or use external parser for Python versions in Poetry specification.
+    # See for details: https://github.com/thoth-station/micropipenv/issues/187
+    wanted_python_version = poetry_lock["metadata"]["python-versions"]
+    message = (
+        "Warning: Currently, Micropipenv is not able to parse complex Python version specifications used by Poetry. "
+        f"Desired version: {wanted_python_version}, current version: {current_python}."
+    )
+    level = "debug" if debug else "warning"
+    getattr(_LOGGER, level)(message)
+
+
+def verify_poetry_lockfile(
+    pyproject=None, poetry_lock=None, current_python_version=None
+):  # type: (Optional[MutableMapping[str, Any]], Optional[MutableMapping[str, Any]], Optional[str]) -> None
+    """Validate that Poetry.lock is up to date with pyproject.toml."""
+    if pyproject is None or poetry_lock is None:
+        poetry_lock, pyproject = _read_poetry()
+    if current_python_version is None:
+        current_python_version = _get_installed_python_version()
+    _validate_poetry_python_version(poetry_lock, current_python_version)
+    pyproject_hash = _compute_poetry_hash(pyproject)
+    poetry_lock_hash = poetry_lock["metadata"]["content-hash"]
+
+    if pyproject_hash != poetry_lock_hash:
+        raise HashMismatch(
+            "Poetry.lock hash {!r} does not correspond to hash computed based on "
+            "pyproject.toml {!r}, aborting deployment".format(poetry_lock_hash, pyproject_hash)
+        )
+
+
+def verify_pipenv_lockfile(
+    pipfile=None, pipfile_lock=None
+):  # type: (Optional[Dict[str, Any]], Optional[Dict[str, Any]]) -> None
+    """Validate that Pipfile.lock is up to date with Pipfile."""
+    pipfile_lock = pipfile_lock or _read_pipfile_lock()
+    python_version = pipfile_lock["_meta"].get("requires", {}).get("python_version")
+    if python_version is not None:
+        if python_version != _get_installed_python_version():
```

Comment:
Nitpick: we can save the result of `_get_installed_python_version()` and use it in the error message below instead of `sys.version_info`.